E2 Applyforce Issue

TheMerkyShadow opened the pull request:
I have no idea why this isn't added.
Without this modification all players can use applyforce by default through e2.
This is extremely dangerous for some servers like mine.
But i really think you guys should add this or something like this. 

Code works as is. Do what you want. I copied the format of propcore.

TheMerkyShadow commented:
I dont why the build failed. lol

TheMerkyShadow commented:
Ok after some careful modifications it is done.
The build is failing but it works fine for me. 
Had to purge old applyforce.
Had to add bone applyforce.

This is finished now.

CaptainPRICE commented:
It is Travis-CI. Due #1369 PR hasn't merged yet, so all PRs are failing, just ignore it..

Divran commented:
Wiremod was made for sandbox, and in sandbox, the ability to apply force to your own entities is not overpowered or dangerous. You can already more or less do that using normal thrusters or wire thruster.
If you're not using wiremod in the environment it was created for, that's not our problem.
You already have a way to block the functions you want blocked anyway, by finding them in the table where they are stored and deleting them or whatever.

I've also thought about a similar thing before. A much more general solution which would allow server owners to allow or deny each individual function. At the time, the original author of E2 (Syranide) shot that idea down with the reasoning that each server should have the same base functions, so that if you make an E2 using these base functions, you can expect it to work on every server. That's also a very good argument.

CaptainPRICE commented on lua/entities/gmod_wire_expression2/core/custom/applyforce.lua:
```diff
@@ -0,0 +1,196 @@
+
+local sbox_E2_ApplyForce = CreateConVar( "sbox_E2_ApplyForce", "1", FCVAR_ARCHIVE ) -- 1: Everyone, 2: Only Admins, 3: Only Superadmins
+
+function ApplyForceValidAction()
```
This function does not belong in here.
Also, cons: do use hook library.

Divran replied:
It does belong there, it just needs to be local.
The hook library has nothing to do with this.

CaptainPRICE replied:
And if he does make if "local", as you say, then he will need to DUPLICATE it, because he also uses it in gates... So, imo, it doesn't belong here.

CaptainPRICE replied:
> The hook library has nothing to do with this.

A lot of Wiremod code lacks hooks. Please start using hooks to allow more control for 3rd party addons! It is my cons, but please do. Thanks.

TheMerkyShadow replied:
Yea i made a mistake wiith the function let me fix it.

TheMerkyShadow commented:
Guys I fixed the local function issue on both e2 and gate side.
Divran I understand you dont want to fix something that isnt broken.
However, it sorta is many server owners lack the understanding and commitment
of disabling applyforce. Some servers have to block e2 altogether because of this.
Espically darkrp servers. Im not asking you to disable anything i reverted some code to allow it by default. However just like holograms and propcore the servers should have the ability within wiremod to disable these because of how powerful they are. Do not deny that this wouldnt be useful.

TheMerkyShadow commented:
I fixed some of the errors but im still testing...
When im done I will tell you its the final version.

TheMerkyShadow commented:
Ok guys @Divran  @CaptainPRICE 
How bout the code now?
I've tested it thoroughly in single-player.

thegrb93 commented:
I don't like the regression in performance due to the additional checks.

TheMerkyShadow commented:
Ok fine so what about a simple on/off without the admin checks?

thegrb93 commented:
Yeah, I'm good with it being an optional extension.

TheMerkyShadow commented:
@thegrb93 I followed your advice. Its just an optional extension now.

CaptainPRICE commented on lua/entities/gmod_wire_expression2/core/custom/applyforce.lua:
```diff
@@ -0,0 +1,195 @@
+e2function void entity:applyForce(vector force)
+	if not validPhysics(this) then return nil end
+	if not isOwner(self, this) then return nil end
+        if not ApplyForceValidAction() then return nil end
```
That's using space as indentation, off the wire coding style.

CaptainPRICE commented on lua/wire/gates/entity.lua:
```diff
@@ -4,6 +4,10 @@
 
 GateActions("Entity")
 
+local function ApplyForceValidAction()
```
Seriously?!

CaptainPRICE commented:
Things to fix:
- Move the order the things are checked. First check if it is valid (in your case, call `ApplyForceValidAction()` before other checks).
- Do not duplicate code like you just have with your `ApplyForceValidAction()` local function..
- Fix mixed indentation.

CaptainPRICE commented:
Wait, what's the purpose of your `ApplyForceValidAction` function at all?

TheMerkyShadow commented:
@CaptainPRICE 
Divran told me that they should be seperate local functions.
I will fix the above problems.

bigdogmat commented on lua/entities/gmod_wire_expression2/core/custom/applyforce.lua:
```diff
@@ -0,0 +1,194 @@
+
+E2Lib.RegisterExtension(ExtName, true, "Allows E2 chips to applyforce", "Allows E2 chips to applyforce"
+
+function ApplyForceValidAction()
```
This has to be defined before you register the extension, otherwise if it's disabled this function will never be created

TheMerkyShadow replied:
didnt know that thanks

bigdogmat commented:
Considering you're just making it so you can disable the extension, I'd suggest not putting it in the custom folder. Also you forgot to copy over the functions from compat

bigdogmat commented on lua/entities/gmod_wire_expression2/core/custom/applyforce.lua:
```diff
@@ -0,0 +1,196 @@
+hook.Call("ApplyForceValidAction", "E2.ApplyForce", function(self)
```
I'm assuming you meant to use `hook.Add` here?

TheMerkyShadow replied:
yes

bigdogmat commented on lua/entities/gmod_wire_expression2/core/custom/applyforce.lua:
```diff
@@ -0,0 +1,196 @@
+end)
+
+function ApplyForceValidAction(self)
+     return hook.Call( "ApplyForceValidAction", GAMEMODE, self )
```
Use `hook.Run` instead

TheMerkyShadow replied:
yes

bigdogmat commented on lua/wire/gates/entity.lua:
```diff
@@ -17,10 +17,11 @@ GateActions["entity_applyf"] = {
 	inputtypes = { "ENTITY" , "VECTOR" },
 	timed = true,
 	output = function(gate, ent, vec )
+               if not ApplyForceValidAction(gate) then return end
```
You're passing the gate here though in the E2 you're passing the context. I'd say pass the player and the entity being forced.

TheMerkyShadow replied:
got it

TheMerkyShadow commented:
@bigdogmat 
better?

bigdogmat commented on lua/entities/gmod_wire_expression2/core/custom/applyforce.lua:
```diff
@@ -0,0 +1,270 @@
+local check = WireLib.checkForce
+
+e2function void applyForce(vector force)
+       if not ApplyForceValidAction(self,this) then return nil end
```
`this` refers to nothing here, it'd be `self.entity`. Same the 2 below.

bigdogmat commented on lua/entities/gmod_wire_expression2/core/custom/applyforce.lua:
```diff
@@ -0,0 +1,270 @@
+	phys:ApplyForceCenter(Vector(force[1],force[2],force[3]))
+end
+
+e2function void applyOfhfsetForce(vector force, vector position)
```
Typo of `Offset`

TheMerkyShadow replied:
lol that was strange

bigdogmat commented:
Take it out of the custom folder. Also like I said, pass the player who owns the gate/E2, not the context/gate itself.

TheMerkyShadow commented:
@bigdogmat 
I was thinking of that myself.
However the player can do that themselves witin the hook with self.

bigdogmat commented:
They can, but it'd require that they can tell the different between the two. If I wanted the player I'd have to do something along the lines of 

```lua
-- Can't think of the wire method to get the owner atm
local owner = isentity(arg1) and arg1:GetCPPIOwner() or arg1.player
```

Plus I don't really see any other useful information you could get out of either of them.

TheMerkyShadow commented:
might be useful somehow.
I will add it.

CaptainPRICE commented on lua/wire/gates/entity.lua:
```diff
@@ -18,9 +18,10 @@ GateActions["entity_applyf"] = {
 		local phys = ent:GetPhysicsObject()
 		if not IsValid( phys ) then return end
-		if not isAllowed( gate, ent ) then return end
+		if not isAllowed( gate, ent ) then return end	
```
Fix your editor preferences.

CaptainPRICE commented on lua/wire/gates/entity.lua:
```diff
@@ -37,8 +38,9 @@ GateActions["entity_applyof"] = {
-	output = function(gate, ent, vec, offset )
+	output = function(gate, ent, vec, offset )             
 		if not IsValid( ent ) then return end
+               if not ApplyForceValidAction(gate,ent) then return end	
```
Fix your editor, use tabs, not spaces.

TheMerkyShadow replied:
ok

TheMerkyShadow commented:
it look okay? The last thing is moving it out of custom folder

Divran commented:
>@CaptainPRICE
Divran told me that they should be seperate local functions.
I will fix the above problems.

no, a minute later I edited my comment

TheMerkyShadow commented:
@Divran 
Im trying to help wiremod and other server owners.
Even if its not merged people can still use my forked version.
